packages/contracts-bedrock/src/L1/OPStackManager.sol

```diff
@@ -0,0 +1,401 @@
+    AddressManager public immutable referenceAddressManager;
+
+    /// @notice The owner of the ProxyAdmin for the reference OP chain.
+    address public immutable referenceProxyAdminOwner;
+
+    /// @notice The reference OP SuperchainConfig used by all chains deployed by this contract.
+    SuperchainConfig public immutable referenceSuperchainConfig;
+
+    /// @notice The reference OP ProtocolVersions used by all chains deployed by this contract.
+    ProtocolVersions public immutable referenceProtocolVersions;
+
+    /// @notice Used to store the addresses of the deployed proxies for a chain,
+    /// to mitigate stack too deep errors.
+    struct Proxies {
+        address l1ERC721Bridge;
+        address l2OutputOracle;
+        address optimismPortal;
+        address systemConfig;
+        address optimismMintableERC20Factory;
+        address l1CrossDomainMessenger;
+        address l1StandardBridge;
+    }
+
+    /// @notice Inputs required to initialize the SystemConfig contract for a new chain.
+    struct SystemConfigInputs {
+        address systemConfigOwner;
+        uint256 overhead;
+        uint256 scalar;
+        bytes32 batcherHash;
+        address unsafeBlockSigner;
+    }
+
+    /// @notice Inputs required to initialize the L2OutputOracle contract for a new chain.
+    struct L2OutputOracleInputs {
+        uint256 submissionInterval;
+        address proposer;
+        address challenger;
+    }
+
+    /// @notice The logic address and initializer selector for an implementation contract.
+    struct Implementation {
+        address logic; // Address containing the deployed logic contract.
+        bytes4 initializer; // Function selector for the initializer.
+    }
+
+    /// @notice Used to set the implementation for a contract by mapping a contract
+    /// name to the implementation data.
+    struct ImplementationSetter {
+        bytes32 name; // Contract name.
+        Implementation info; // Implementation to set.
+    }
+
+    /// @notice Returns the latest approved release of the OP Stack contracts.
+    /// Release strings follow semver and are named with the format `op-contracts/vX.Y.Z`.
+    bytes32 public latestRelease;
+
+    /// @notice Maps a release version to a contract name to it's implementation data.
+    mapping(bytes32 => mapping(bytes32 => Implementation)) public implementations;
+
+    /// @notice Maps an L2 Chain ID to the SystemConfig for that chain.
+    /// Most information for a chain can be found from it's the SystemConfig, with ProtocolVersions
+    /// being the exception. ProtocolVersions is a singleton for all chains deployed by this
+    /// contract, so `referenceProtocolVersions` is used for all chains.
+    /// Additionally, the ProxyAdmin can be discovered offchain, but cannot be discovered onchain
+    /// due to implementation details of the ProxyAdmin contract.
+    mapping(uint256 => SystemConfig) public systemConfigs;
```

Comment:
We may not want to store this on-chain. Is it possible to just pass in whatever systemConfig address is needed?